src/jdk.incubator.code/share/classes/jdk/incubator/code/interpreter/Interpreter.java

```diff
@@ -100,7 +101,27 @@ Object invoke(MethodHandles.Lookup l, T op,
                     String.format("Actual #arguments (%d) differs from #parameters (%d) plus #captured arguments (%d)",
                             args.size(), parameters.size(), capturedValues.size())));
         }
-
+        // validate runtime args types
+        List<Value> symbolicValues = Stream.concat(parameters.stream(), capturedValues.stream()).toList();
+        for (int i = 0; i < symbolicValues.size(); i++) {
+            Value sv = symbolicValues.get(i);
+            Object rv = args.get(i);
+            try {
+                JavaType typeToResolve = switch (sv.type()) {
+                    case PrimitiveType pt -> pt.box().orElseThrow();
+                    case JavaType jt -> jt;
+                    default -> throw new IllegalStateException("Unexpected value: " + sv.type());
+                };
+                Class<?> c = typeToResolve.toNominalDescriptor().resolveConstantDesc(l);
+                if (!c.isInstance(rv)) {
+                    Class<?> svc = ((JavaType) sv.type()).toNominalDescriptor().resolveConstantDesc(l);
```

Comment:
You don't have to resolve here, we can just directly use `sv.type()` as an argument for the exception message.